### PR TITLE
Fulltext searching in datastore query

### PR DESCRIPTION
**Pull request by dafeder**
This is a somewhat experimental feature for now and will not be fully documented until we have used it a bit more in the wild. But this allows fulltext boolean match queries against a datastore resource in MySQL if the column(s) being queries are properly indexed.

This PR also adds two new keys to the schema object in the datastore response. One is `indexes` which lists all standard indexes on the database table, as per the Drupal Schema API spec. Another is `fulltext indexes`, which follows the same pattern. This is not part of the official Drupal Schema spec but I felt was needed to signal to the user what columns are available for fulltext searching, and matches the rest of the spec as closely as possible.

## QA steps

This will need to be test locally since for now we need to create the fulltext indexes manually. Set up a local site with the demo content.

1. Ensure that you have a datastore at /api/1/datastore/query/d460252e-d42c-474a-9ea9-5287b1d595f6/0
2. Try running a query before indexing. You should get a 400 error "You have attempted a fulltext match against a column that is not indexed for fulltext searching."
```http
POST https://fulltext-query.ddev.site/api/1/datastore/query/d460252e-d42c-474a-9ea9-5287b1d595f6/0

{
  "conditions": [
    {
      "property": "city",
      "operator": "match",
      "value": "York"
    }
  ]
}
```
3. Use `drush dkan:dataset-info d460252e-d42c-474a-9ea9-5287b1d595f6` to get the table name of the datastore. Let's say its `datastore_efee6d8c0ee35d0b1a161743806d018d`.
4. Connect to your db CLI with `drush sqlc`.
5. Create a fulltext index on the city field: 
```sql
ALTER TABLE datastore_efee6d8c0ee35d0b1a161743806d018d ADD FULLTEXT INDEX `cityfulltext` (`city`);
```
6. Retry the query from step 2.
7. Confirm that you see a result and got a single hit for "New York."
8. Try again with a wildcard:
```http
POST https://fulltext-query.ddev.site/api/1/datastore/query/d460252e-d42c-474a-9ea9-5287b1d595f6/0

{
  "conditions": [
    {
      "property": "city",
      "operator": "match",
      "value": "Yor*"
    }
  ]
}
```
9. Confirm that you still get a hit for "New York".
10. Check the `schema` object in the response and confirm that the fulltext index is listed properly.
11. Now let's try an index with two columns. Go back to the db CLI and let's create a second index that combines city and state:
```sql
ALTER TABLE datastore_efee6d8c0ee35d0b1a161743806d018d ADD FULLTEXT INDEX `citystate` (`city`, `state`);
```
12. We can try a few queries that should demonstrate how this can be leveraged:
```http
POST https://fulltext-query.ddev.site/api/1/datastore/query/d460252e-d42c-474a-9ea9-5287b1d595f6/0

{
  "conditions": [
    {
      "property": "city,state",
      "operator": "match",
      "value": "San"
    }
  ]
}
```
```http
POST https://fulltext-query.ddev.site/api/1/datastore/query/d460252e-d42c-474a-9ea9-5287b1d595f6/0

{
  "conditions": [
    {
      "property": "city,state",
      "operator": "match",
      "value": "Texas"
    }
  ]
}
```
13. Confirm that in both cases you get multiple hits from both city and state fields. Also confirm that the newer fulltext index is included correctly in the schema object.

### CODE REVIEW

**Comment by dafeder** [comment]
CodeClimate is complaining about 21 methods in SelectFactory. I think a refactor there that would actually lead to cleaner code would be a big lift and make more sense as part of work to pull AbstractDatbaseTable and its derivatives out of the rest of DKAN. That's way out of scope for this issue so I'd argue that we should mark this as a known issue for now but give it a pass. I'll leave that up to whoever reviews this though.

File: modules/datastore/tests/src/Unit/Storage/DatabaseTableTest.php
Comment (clayliddell) [inline modules/datastore/tests/src/Unit/Storage/DatabaseTableTest.php:82]: Could we have one "indexes" array, similar to what we're planning to do for data-dictionaries?
e.g.
```
"indexes" => [
  "index_name" => {
    "fields" => [
      "first_name",
      "last_name",
    ],
    "type" => "fulltext" or "btree" or "spatial",
  },
]
```

File: modules/datastore/src/Storage/DatabaseTable.php
Comment (clayliddell) [inline modules/datastore/src/Storage/DatabaseTable.php]: just adding a return type
```suggestion
  protected function addIndexInfo(array &$schema): void {
```

File: modules/datastore/src/Storage/DatabaseTable.php
Comment (clayliddell) [inline modules/datastore/src/Storage/DatabaseTable.php:189]: just adding a return type
```suggestion
  protected function buildTableSchema(string $tableName, array $fieldsInfo): array {
```

File: modules/datastore/tests/src/Unit/Storage/DatabaseTableTest.php
Comment (clayliddell) [inline modules/datastore/tests/src/Unit/Storage/DatabaseTableTest.php:440]: I think we would benefit from having a integration/functional test here to ensure the underlying MySQL error doesn't change in future versions.

File: modules/common/tests/src/Unit/Storage/QueryDataProvider.php
Comment (clayliddell) [inline modules/common/tests/src/Unit/Storage/QueryDataProvider.php:383]: I don't think there's currently a test for this method.

**Review by clayliddell — COMMENTED** [review]
The overall structure looks good. I've left a couple suggestions related to test coverage and one comment related to the new index portion of the schema.

File: modules/datastore/tests/src/Unit/Storage/DatabaseTableTest.php
Comment (dafeder) [reply modules/datastore/tests/src/Unit/Storage/DatabaseTableTest.php:82]: That is what I think we should do if we switch to Frictionless' table schema for the schema object. For better or worse we are still using the [Drupal schema API](https://api.drupal.org/api/drupal/core!lib!Drupal!Core!Database!database.api.php/group/schemaapi/9.0.x) here.

File: modules/common/tests/src/Unit/Storage/QueryDataProvider.php
Comment (dafeder) [reply modules/common/tests/src/Unit/Storage/QueryDataProvider.php:383]: I'm confused... this method is part of a test

File: modules/datastore/tests/src/Unit/Storage/DatabaseTableTest.php
Comment (dafeder) [reply modules/datastore/tests/src/Unit/Storage/DatabaseTableTest.php:440]: I see your point... FWIW the only part being matched is "Can't find FULLTEXT index matching the column list" which I think is unlikely to change anytime soon. And if it did it would just fall back to our generic error message which wouldn't be a big deal. 

I'm hesitant because we don't actually have any functional tests for the datastore right now and there's a fairly high LOE for adding that into this PR, and I don't know that this particular thing justifies doing that. We should reexamine testing for the datastore API but IMO that's out of scope for this ticket. We can discuss further if you feel strongly though.

**Comment by dafeder** [comment]
Whoops somehow merging those suggestions broke everything @clayliddell - will investigate. FWIW I'm not a big fan of that feature in github. Also I noticed a potential issue in my code - I just have that one placeholder `:words` for the matching terms, which will cause an issue if someone adds multiple MATCH conditions to a single query. I'll add an iterator to the end of it so we get `:words0` `words1` etc.

File: modules/datastore/tests/src/Unit/Storage/DatabaseTableTest.php
Comment (clayliddell) [reply modules/datastore/tests/src/Unit/Storage/DatabaseTableTest.php:440]: No, I kind of agree - we should hold off for now, at least pending datastore refactor efforts

File: modules/datastore/src/Storage/DatabaseTable.php
Comment (clayliddell) [inline modules/datastore/src/Storage/DatabaseTable.php]: I believe this is the one causing the problem:
```suggestion
  protected function buildTableSchema(string $tableName, array $fieldsInfo) {
```

File: modules/datastore/tests/src/Unit/Storage/DatabaseTableTest.php
Comment (clayliddell) [reply modules/datastore/tests/src/Unit/Storage/DatabaseTableTest.php:82]: I'm fine with just sticking to the Drupal schema for now - may be worth revisiting later though

**Review by clayliddell — APPROVED** [review]

**Pull request merged**
